File: src/accessiweather/weather_history.py
"""
Weather history comparison functionality using Open-Meteo archive API.

This module provides functionality to compare current weather conditions with
historical data from Open-Meteo's archive endpoint. No local storage required.
Designed with accessibility in mind for screen reader users.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from datetime import date, datetime, timedelta
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .models import CurrentConditions, Location

logger = logging.getLogger(__name__)


@dataclass
class HistoricalWeatherData:
    """Historical weather data for a specific date."""

    date: date
    temperature_max: float
    temperature_min: float
    temperature_mean: float
    condition: str
    humidity: int | None
    wind_speed: float
    wind_direction: int | None
    pressure: float | None


@dataclass
class WeatherComparison:
    """Comparison between current weather and historical data."""

    temperature_difference: float
    temperature_description: str
    condition_changed: bool
    previous_condition: str
    condition_description: str | None
    days_ago: int

    @classmethod
    def compare(
        cls,
        current: CurrentConditions,
        historical: HistoricalWeatherData,
        days_ago: int,
    ) -> WeatherComparison:
        """
        Compare current conditions with historical data.

        Args:
            current: Current weather conditions
            historical: Historical weather data
            days_ago: Number of days in the past

        Returns:
            A WeatherComparison instance with comparison details

        """
        # Use mean temperature for comparison
        temp_diff = current.temperature - historical.temperature_mean

        # Generate temperature description
        if abs(temp_diff) < 1.0:
            temp_desc = "about the same temperature"
        elif temp_diff > 0:
            temp_desc = f"{abs(temp_diff):.1f} degrees warmer"
        else:
            temp_desc = f"{abs(temp_diff):.1f} degrees cooler"

        # Check condition change
        condition_changed = current.condition != historical.condition
        condition_desc = None
        if condition_changed:
            condition_desc = f"Changed from {historical.condition} to {current.condition}"

        return cls(
            temperature_difference=temp_diff,
            temperature_description=temp_desc,
            condition_changed=condition_changed,
            previous_condition=historical.condition,
            condition_description=condition_desc,
            days_ago=days_ago,
        )

    def get_accessible_summary(self) -> str:
        """
        Generate a screen-reader friendly summary of the comparison.

        Returns:
            A human-readable summary of weather changes

        """
        parts = []

        # Temperature summary
        if self.days_ago == 1:
            time_ref = "yesterday"
        elif self.days_ago == 7:
            time_ref = "last week"
        else:
            time_ref = f"{self.days_ago} days ago"

        parts.append(f"Compared to {time_ref}: {self.temperature_description}")

        # Condition summary
        if self.condition_changed and self.condition_description:
            parts.append(self.condition_description)

        return ". ".join(parts) + "."


class WeatherHistoryService:
    """Service for fetching historical weather data and making comparisons."""

    def __init__(self, openmeteo_client=None):
        """
        Initialize the weather history service.

        Args:
            openmeteo_client: Optional OpenMeteoApiClient instance. If not provided,
                            one will be created.

        """
        if openmeteo_client is None:
            from .openmeteo_client import OpenMeteoApiClient

            self.openmeteo_client = OpenMeteoApiClient(
                user_agent="AccessiWeather/2.0",
                timeout=30.0,
            )
        else:
            self.openmeteo_client = openmeteo_client

    def get_historical_weather(
        self,
        latitude: float,
        longitude: float,
        target_date: date,
        temperature_unit: str = "fahrenheit",
    ) -> HistoricalWeatherData | None:
        """
        Fetch historical weather data for a specific date.

        Args:
            latitude: Latitude of the location
            longitude: Longitude of the location
            target_date: Date to fetch historical data for
            temperature_unit: Temperature unit ("celsius" or "fahrenheit")

        Returns:
            HistoricalWeatherData if successful, None otherwise

        """
        try:
            # Use Open-Meteo archive endpoint for historical data
            params = {
                "latitude": latitude,
                "longitude": longitude,
                "start_date": target_date.isoformat(),
                "end_date": target_date.isoformat(),
                "daily": [
                    "weather_code",
                    "temperature_2m_max",
                    "temperature_2m_min",
                    "temperature_2m_mean",
                    "wind_speed_10m_max",
                    "wind_direction_10m_dominant",
                ],
                "temperature_unit": temperature_unit,
                "timezone": "auto",
            }

            # Call archive endpoint (different from forecast)
            response = self.openmeteo_client._make_request("archive", params)

            if not response or "daily" not in response:
                logger.warning(f"No historical data available for {target_date}")
                return None

            daily = response["daily"]

            # Extract data for the requested date
            if not daily.get("time") or len(daily["time"]) == 0:
                return None

            # Get weather description from code
            weather_code = daily.get("weather_code", [0])[0]
            condition = self.openmeteo_client.get_weather_description(weather_code)

            return HistoricalWeatherData(
                date=target_date,
                temperature_max=daily.get("temperature_2m_max", [0])[0],
                temperature_min=daily.get("temperature_2m_min", [0])[0],
                temperature_mean=daily.get("temperature_2m_mean", [0])[0],
                condition=condition,
                humidity=None,  # Not available in archive endpoint
                wind_speed=daily.get("wind_speed_10m_max", [0])[0],
                wind_direction=daily.get("wind_direction_10m_dominant", [None])[0],
                pressure=None,  # Not available in archive endpoint
            )

        except Exception as e:
            logger.error(f"Failed to fetch historical weather data: {e}")
            return None

    def compare_with_yesterday(
        self,
        location: Location,
        current_conditions: CurrentConditions,
        temperature_unit: str = "fahrenheit",
    ) -> WeatherComparison | None:
        """
        Compare current weather with yesterday's weather.

        Args:
            location: Location for the comparison
            current_conditions: Current weather conditions
            temperature_unit: Temperature unit for API request

        Returns:
            WeatherComparison if historical data is available, None otherwise

        """
        yesterday = (datetime.now() - timedelta(days=1)).date()
        historical = self.get_historical_weather(
            location.latitude, location.longitude, yesterday, temperature_unit
        )

        if historical is None:
            return None

        return WeatherComparison.compare(current_conditions, historical, days_ago=1)

    def compare_with_last_week(
        self,
        location: Location,
        current_conditions: CurrentConditions,
        temperature_unit: str = "fahrenheit",
    ) -> WeatherComparison | None:
        """
        Compare current weather with weather from one week ago.

        Args:
            location: Location for the comparison
            current_conditions: Current weather conditions
            temperature_unit: Temperature unit for API request

        Returns:
            WeatherComparison if historical data is available, None otherwise

        """
        last_week = (datetime.now() - timedelta(days=7)).date()
        historical = self.get_historical_weather(
            location.latitude, location.longitude, last_week, temperature_unit
        )

        if historical is None:
            return None

        return WeatherComparison.compare(current_conditions, historical, days_ago=7)

    def compare_with_date(
        self,
        location: Location,
        current_conditions: CurrentConditions,
        target_date: date,
        temperature_unit: str = "fahrenheit",
    ) -> WeatherComparison | None:
        """
        Compare current weather with weather from a specific date.

        Args:
            location: Location for the comparison
            current_conditions: Current weather conditions
            target_date: Date to compare with
            temperature_unit: Temperature unit for API request

        Returns:
            WeatherComparison if historical data is available, None otherwise

        """
        days_ago = (datetime.now().date() - target_date).days
        historical = self.get_historical_weather(
            location.latitude, location.longitude, target_date, temperature_unit
        )

        if historical is None:
            return None

        return WeatherComparison.compare(current_conditions, historical, days_ago=days_ago)

File: rust/crates/aw-core/src/weather_history.rs
//! Historical weather comparison. Pure part of
//! `accessiweather/weather_history.py`; the archive fetch lives in
//! `aw_providers::client::history`.

use chrono::NaiveDate;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::model::CurrentConditions;

/// Daily fields requested from the Open-Meteo archive.
pub const HISTORY_DAILY_FIELDS: [&str; 6] = [
    "weather_code",
    "temperature_2m_max",
    "temperature_2m_min",
    "temperature_2m_mean",
    "wind_speed_10m_max",
    "wind_direction_10m_dominant",
];

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HistoricalWeatherData {
    pub date: NaiveDate,
    pub temperature_max: f64,
    pub temperature_min: f64,
    pub temperature_mean: f64,
    pub condition: String,
    pub humidity: Option<i64>,
    pub wind_speed: f64,
    pub wind_direction: Option<i64>,
    pub pressure: Option<f64>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WeatherComparison {
    pub temperature_difference: f64,
    pub temperature_description: String,
    pub condition_changed: bool,
    pub previous_condition: String,
    pub condition_description: Option<String>,
    pub days_ago: i64,
}

impl WeatherComparison {
    /// Compare current conditions with a historical day's mean temperature.
    /// `None` when the current temperature is unknown (Python raises there).
    pub fn compare(
        current: &CurrentConditions,
        historical: &HistoricalWeatherData,
        days_ago: i64,
    ) -> Option<Self> {
        let diff = current.temperature? - historical.temperature_mean;
        let temperature_description = if diff.abs() < 1.0 {
            "about the same temperature".to_string()
        } else if diff > 0.0 {
            format!("{:.1} degrees warmer", diff.abs())
        } else {
            format!("{:.1} degrees cooler", diff.abs())
        };
        let condition_changed = current.condition.as_deref() != Some(historical.condition.as_str());
        let condition_description = condition_changed.then(|| {
            format!(
                "Changed from {} to {}",
                historical.condition,
                current.condition.as_deref().unwrap_or("None")
            )
        });
        Some(Self {
            temperature_difference: diff,
            temperature_description,
            condition_changed,
            previous_condition: historical.condition.clone(),
            condition_description,
            days_ago,
        })
    }

    /// `get_accessible_summary`.
    pub fn accessible_summary(&self) -> String {
        let time_ref = match self.days_ago {
            1 => "yesterday".to_string(),
            7 => "last week".to_string(),
            n => format!("{n} days ago"),
        };
        let mut parts = vec![format!(
            "Compared to {time_ref}: {}",
            self.temperature_description
        )];
        if self.condition_changed {
            if let Some(desc) = &self.condition_description {
                parts.push(desc.clone());
            }
        }
        parts.join(". ") + "."
    }
}

fn first_value<'a>(daily: &'a Value, field: &str) -> Option<&'a Value> {
    daily
        .get(field)?
        .as_array()?
        .first()
        .filter(|v| !v.is_null())
}

fn as_float(value: &Value) -> Option<f64> {
    match value {
        Value::Number(n) => n.as_f64(),
        Value::String(s) => s.trim().parse().ok(),
        _ => None,
    }
}

fn as_int(value: &Value) -> Option<i64> {
    match value {
        Value::Number(n) => n.as_i64().or(n.as_f64().map(|f| f.trunc() as i64)),
        Value::String(s) => s.trim().parse().ok(),
        _ => None,
    }
}

/// Parse a single-day archive response (`get_historical_weather`).
/// `describe` maps the WMO weather code to text (Open-Meteo's
/// `get_weather_description`).
pub fn parse_historical_weather(
    response: &Value,
    target_date: NaiveDate,
    describe: impl Fn(&Value) -> String,
) -> Option<HistoricalWeatherData> {
    let daily = response.get("daily")?;
    if daily.get("time")?.as_array()?.is_empty() {
        return None;
    }
    let weather_code = first_value(daily, "weather_code")?;
    let temperature_max = as_float(first_value(daily, "temperature_2m_max")?)?;
    let temperature_min = as_float(first_value(daily, "temperature_2m_min")?)?;
    let temperature_mean = as_float(first_value(daily, "temperature_2m_mean")?)?;
    let wind_speed = as_float(first_value(daily, "wind_speed_10m_max")?)?;
    let wind_direction = match first_value(daily, "wind_direction_10m_dominant") {
        Some(v) => Some(as_int(v)?),
        None => None,
    };
    Some(HistoricalWeatherData {
        date: target_date,
        temperature_max,
        temperature_min,
        temperature_mean,
        condition: describe(weather_code),
        humidity: None,
        wind_speed,
        wind_direction,
        pressure: None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::golden::{self, field};

    #[test]
    fn golden_comparisons() {
        let cases = golden::load("history/comparisons.json");
        for case in cases.as_array().unwrap() {
            let current: CurrentConditions = field(case, "current");
            let historical: HistoricalWeatherData = field(case, "historical");
            let comparison =
                WeatherComparison::compare(&current, &historical, field(case, "days_ago"));
            let expected: WeatherComparison = field(case, "comparison");
            assert_eq!(comparison.as_ref(), Some(&expected), "{}", case["name"]);
            assert_eq!(
                comparison.unwrap().accessible_summary(),
                case["summary"].as_str().unwrap()
            );
        }
    }

    #[test]
    fn golden_archive_parsing() {
        let cases = golden::load("history/archive_parsing.json");
        for case in cases.as_array().unwrap() {
            let date: NaiveDate = field(case, "date");
            let parsed =
                parse_historical_weather(&case["response"], date, |code| format!("code {code}"));
            let expected: Option<HistoricalWeatherData> = field(case, "result");
            assert_eq!(parsed, expected, "{}", case["name"]);
        }
    }
}
